docs/adr/000x-modelling-history.md

```diff
@@ -0,0 +1,314 @@
+and live, we can only know about one of these at any time and not of any
+previous versions.
+
+The suggestion is that we enhance Content Publisher to support the concepts of
+Document, Edition and Revision.
+
+For a document with a live version and a draft this could enable:
+
+- The ability to discard the draft and revert the current version in Content
+  Publisher back to the live version (including all images and attachments) to
+  be in sync with Publishing API;
+- The ability to have the details of the live version, most noticeably state
+  and URL, so this can be presented to users;
+- Ability to re-present either the live or draft version to Publishing API
+  should there be any problems with publishing.
+
+And with regards to history this could enable:
+
+- A record of who did what to a document and when - effectively an audit trail
+  of accountability;
+- Ability to compare two revisions of a document to show changes;
+- Storing a full history of a document - to enable importing Whitehall
+  documents with histories pre-dating the Publishing API.
+
+A [MoSCoW](https://en.wikipedia.org/wiki/MoSCoW_method) breakdown of this
+functionality is defined as:
+
+### Must
+
+- Store full information on current and live versions of a document;
+  *because* we may need to re-present either to the Publishing API as Content
+  Publisher is the source of truth.
+- Consider how a revision of a document and aspects of it have state that may
+  change;
+  *because* the concept of revision could become convoluted if it stores
+  information that a user does not directly impact.
+
+### Should
+
+- Have the full information for every edition that was published on GOV.UK;
+  *because* this will allow us the means to present what changed between
+  published editions - which is needed for a variety reasons, including, in
+  some cases, legal.
+- Record every change a publisher makes and allow for that information to be
+  accessible;
+  *because* this maintains options for TimelineEntry and provides rich
+  debugging information.
+- Avoid relying on the TimelineEntry model as the main answer to any
+  historical queries;
+  *because* this model is close to the user and is likely to be iterated. The
+  deeper the coupling to a document the more difficult this becomes.
+
+
+### Could
+
+- Link TimelineEntry to changes rather than storing the information twice;
+  *because* this would reduce concerns and opportunities for error.
+- Consider states to be something that may change in a background job;
+  *because* publishing can be a slow task that could timeout a single
+  HTTP request, considering this up-front will be less painful than bolting it
+  on.
+- Enable a future feature to enable reverting to different points in time;
+  *because* this could enable users to make minor changes to a live document
+  and then recover a draft.
+- Separate related data to different tables to reduce data usage;
```

Comment:
Sure, this'll be another area where I struggled to be succinct.

It's based on the idea of storing bits of revisions in different tables that are unlikely to change at the same time so that a revision itself isn't a huge amount of data.

I'll have a think about better ways to phrase it.

Reply:
I've dropped this section now